### ISSUE TITLE
Prevent generation of invalid identifiers

### ISSUE THREAD

**Issue by moltob** [pyecore/pyecoregen#2]
I am using the generator to create Python classes from a metamodel for [Franca IDL](https://github.com/franca/franca). The ecore model defined by the project contains some classes used in an expression language, e.g. there is a meta-class ``else``. Right now, pyecoregen will therefore produce invalid Python code since Python keywords must not be used as identifiers.

This enhancement will add a Jinja filter to ensure conflicting names are adapted during code generation. The Pythonic way of adaptation is to add a trailing underscore, which is what this filter would be doing.

**Comment by moltob** [comment]
@aranega Please comment if you have feedback. I will do this through a PR for informational purpose.

**Comment by aranega** [comment]
Good catch! I totally forgot about reserved keywords. The only question I have is related to the best translation. I see two ways of translating this regarding your comment. Each solution impacts introspection and I cannot see which one is the best:

```Python
# First solution
class A(EObject, metaclass=MetaEClass):
    else_ = EAttribute(eType=EString)


a = A()
a.eClass.findEStructuralFeature('else')  # returns None
attribute = a.eClass.findEStructuralFeature('else_')  # need to use the '_'

# but
getattr(a, attribute.name)  # works
```
And the second solution:

```Python
class A(EObject, metaclass=MetaEClass):
    else_ = EAttribute(name='else', eType=EString)   # we change the name


a = A()
attribute = a.eClass.findEStructuralFeature('else')  # returns the right EAttribute

# but
getattr(a, attribute.name)  # does not work as entended (EValue is reached)
```
I would be slightly in favor of the first one, but I'm not certain. Which one sounds the best to you?

**Comment by moltob** [comment]
I favor the first one, which to me appears to be a consistent, no-nonsense translation. :-) I would output a warning via logger so users can take notice.

**Comment by aranega** [comment]
The warning log is a really great with the first solution! I will also update the Acceleo transformation to deal with those cases, but I'm pretty sure it will be way more verbose that your solution ;)

**Comment by moltob** [comment]
I started an implementation with the filters but this quickly becomes awkward, since I have to introduce the filter in _many_ places in the templates and additionally, some of my utility filters (like ``supertypes``) also access names in Python code, i.e. there the translation would be required as well.

While I will write tests, I feel this is fragile in terms of having forgotten something, and also against future extensions ("did I add the name translation to all new code?"). Therefore I will do it differently, at the model level, essentially replacing bad names right there, before generation even starts.

To do that, I see different options:

* Walk the model and update names in place. Note that this will modify the user's model, which I don't really want to do.
* Walk and copy the model via reflection and essentially update all name properties in my copy. For large models, this may become an issue at some point.
* Monkey-patch the name accessor of ``NamedElement`` to do what I want. But that is a bit rude...

Do you see other options? Does pyecore have a walker or visitor already somewhere that I could use? Or a mechanism to clone the model? Or would you continue with the original approach?

I'd appreciate your input.

PS: I'll be on vacation starting tomorrow for two weeks (in France actually) and I have been told to leave my laptop at home... :-) I'll respond to comments but code won't come during that time.

**Comment by aranega** [comment]
Outch, that's a tough one. In the Acceleo version, I don't have many option, so I will go for a query that I need to manually call to deal with it. 
In your case, I couldn't tell which solution will react the best. The 1st and 3rd solution would work lovely, but in case the generator is included as part of a code and not used as a 'one shot tool', the modificatins could introduce strange side-effects.
The 2nd solution will indeed have some performance issue at some points. Moreover, I never tried to copy model elements, I know that with cross-references and stuffs, EMF as some issues with that, and I think pyecore might suffer from the same issues (but, again, I never tested anything about it).

So, as long as the monkey patch could be 'undone', it could do the trick, but I'm affraid that it will be difficult to do. The `name` behavior is handled in the `ENamedElement` (simple attribute) and also 'overriden' in the `EStructuralFeature` class as a property. I'm clearly no expert in this, perhaps a custom property could help to deal with all of this?

Also, I could introduce (or could be introduced by monkey patching) a kind of `secure_name` property that compute the right name, but it will be almost the same solution than the one you already implemented, perhaps a little bit less verbose, but you have to think to each locations anyway.

Otherwise, the only 'walker' I have is the `eAllContents()` that returns a generator on every contained element from a point, and I don't have clone method :( .

To sum up:
* there is still the solution you are working with,
* the solution where I had (or monkey patch add) a `secure_name` or `unrestricted_name` property in the `ENamedElement` class,
* the monkey patch solution if you can deal with it and if it could be undone (but the undone part is not absolutely required in the first place, it could be a special case anyway),
* as a last resort, trying to 'clone' the model.

I'm sorry, I now that it probably doesn't help you much, but currently, I don't see other solutions.

Anyway, don't bother a lot with all of this and enjoy your vacations!

**Comment by moltob** [comment]
I played a bit with patching.

Doing it for ``EStructuralFeature`` will work fine, much like a decorator. I can patch it once for the class and remember and delegate to the existing property via my overriding property getter function. However, for ``NamedElement`` this does not work I am afraid. The instance attribute is simply hidden by my property. At least at class level, this is not possible.

What I could do: Temporarily change all affected _instance_ names and reset them afterwards, but that feels wrong.

We could make the name attribute in pyecore a property as well. That would allow patching it at class level. I'll give that a try. The change to pyecore should be minimal.

**Comment by moltob** [comment]
Woa, that breaks a few things. The only thing I did is to move the property based implementation of ``name`` from ``EstructuralFeature`` up to ``ENamedElement``, and only left the ``super()`` initialization call passing the name to ``NamedElement``.

But then pyecore can't even be imported, all UTs fail with:

```
tests\xmi\test_delete.py:2: in <module>
    import pyecore.ecore as Ecore
pyecore\ecore.py:1268: in <module>
    Core.register_classifier(EObject, promote=True)
pyecore\ecore.py:123: in register_classifier
    Core._promote(cls, abstract)
pyecore\ecore.py:118: in _promote
    operation.eParameters.append(parameter)
pyecore\ecore.py:531: in append
    self.add(value, update_opposite)
pyecore\ecore.py:537: in add
    self._update_opposite(value, self._owner)
pyecore\ecore.py:413: in _update_opposite
    object.__getattribute__(owner, eOpposite.name)  # Force load
E   TypeError: attribute name must be string, not 'NoneType'
```

From an outside look it was a bit suspicious that ``NamedElement``'s implementation of the ``name`` attrbite was not OK for ``EStructuralFeature`` to use. Do you have a hint for me what's going on?

**Comment by aranega** [comment]
Indeed, I was afraid of such a thing. 
Well, it is a little bit special, I remember I had an infinite recursion using the default `ENamedElement` for `EStructuralFeature`. The issue was coming from the `__get__` and `__set__` as the `EStructuralFeature` acts like an attribute descriptor. I think the fact that it searches for the `name` attribute at some point, which is described using a `EAttribute`, which is an `EStructuralFeature`, which is a descriptor... was an issue. The introduction of `_name` was stopping the recursion as `_name` is not an `EStructuralFeature` (and so, not a descriptor).

I think the modification is not so trivial after all, I need to go in depth and check how I can modify this. Perhaps, the fact that the core of pyecore uses the classes to build the M2/M3 layer introduces this.

I keep you informed of what I find! Really sorry about that.

**Comment by aranega** [comment]
Sooo, I came accross the code... My bad, it was super easy to fix actually, my mistake was to override the class attribute `name` in `ENamedElement` during `EClasses` registration...

I simply changed the static access from `name` to `name_` and all is good (I just need to patch two tests that does not uses the right names). And everything should be easier to monkey patch (hopefully).

I commit/push and prepare a new pyecore release.

**Comment by moltob** [comment]
Perfect, I'll look at it and give it a try.

**Comment by aranega** [comment]
Sorry Mike, I've just saw your ticket (wasn't home and forgot my phone). I'm looking/patching pyecore following your comments. 
Again, sorry!

**Comment by aranega** [comment]
Ok, so it turns out that I can direcly remove the property, `name` works (and faster) even as a simple attribute. As in your PR, you are overriding `__getattribute__()`, it should work as intended without the use of a decorator, which is great! I'm pushing (again) a new version. Since I know travis deploys for me, I no longer fear to fail a new pypi release.

Thanks for pointing out my crappy solution, your reviews are really gold for me!

**Comment by aranega** [comment]
And merged! I close the issue.
